### DISCUSSION TITLE
NODE_OPTIONS='--inspect' does not work as I expect

### DISCUSSION THREAD

**Discussion by chrisfillmore** [vercel/next.js#46894] [Help]
### Summary

I am running into perhaps two confounding issues. I could use some help diagnosing.

My first issue is this: `NODE_OPTIONS='--inspect' next dev` does not work as I expect

It is perhaps the same as this issue: https://github.com/vercel/next.js/issues/11821

I created a brand new app created with `npx create-next-app`. My `dev` script is as so:
```
"dev": "NODE_OPTIONS='--inspect' next dev"
```

Now if I run this command I get the following output:
```
➜  my-app git:(main) ✗ npm run dev

> my-app@0.1.0 dev
> NODE_OPTIONS='--inspect' next dev

Debugger listening on ws://127.0.0.1:9229/97da19bc-fe13-453c-b7cf-c49a8cf2ce65
For help, see: https://nodejs.org/en/docs/inspector
Starting inspector on 127.0.0.1:9229 failed: address already in use
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
Starting inspector on 127.0.0.1:9229 failed: address already in use
```

if I open a browser to `localhost:3000`, I see the Vercel/NextJS default page.

So `next dev` appears to be starting a debugger on port `9229`. Then tries to do so again and fails. If I open `chrome://inspect`, I can see the `next` application under Remote Target. I click inspect, it opens DevTools, but it is not actually connected to the running `next` application. No logging appears in the console, there are no Sources to inspect. Searching for any files with `cmd+P` does not yield anything relevant, as far as I can tell.

My colleague who is working on a NextJS project with me experiences the same problem. Similar environment setup to mine (see additional info below).

In case this is useful, at this time, if I run `lsof -i :9229`, I get the following output:
```
COMMAND     PID  USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
Google     6293 chris  154u  IPv4 0xb86faa474cbc9969      0t0  TCP localhost:53944->localhost:9229 (ESTABLISHED)
node      19086 chris   26u  IPv4 0xb86faa474cbc3411      0t0  TCP localhost:9229 (LISTEN)
node      19086 chris   32u  IPv4 0xb86faa474c885969      0t0  TCP localhost:9229->localhost:53944 (ESTABLISHED)
```

On the other hand, if I do this:
```
node --inspect ./node_modules/next/dist/bin/next dev
```
I get the following output:
```
➜  my-app git:(main) ✗ node --inspect ./node_modules/next/dist/bin/next dev
Debugger listening on ws://127.0.0.1:9229/ed800546-d6f5-4a4a-8068-42c766b5aaa1
For help, see: https://nodejs.org/en/docs/inspector
warn  - the --inspect option was detected, the Next.js server should be inspected at port 9230.
Debugger listening on ws://127.0.0.1:9230/e5a9e304-1dfa-43dd-8c9a-b9dfec7d8d9d
```

Now if I open `chrome://inspect` (and if I configure `localhost:9230` as a possible target for inspection), I can see a second debug target on port `9230`. If I `inspect` this target, I can see it is attached to my application. Logs appear in the console, and I can `cmd+P` to search for my application. I can set breakpoints and hit them.

So my first question is: should I expect `next` to start two debuggers like this? What is the expected behaviour?

In case it is relevant, at this stage if I run `lsof -i`:
```
$ lsof -i :9229
COMMAND   PID  USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
node    19206 chris   26u  IPv4 0xb86faa474cbd6969      0t0  TCP localhost:9229 (LISTEN)

$ lsof -i :9230
COMMAND     PID  USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
Google     6293 chris  154u  IPv4 0xb86faa474c8cb969      0t0  TCP localhost:63416->localhost:9230 (ESTABLISHED)
node      19207 chris   26u  IPv4 0xb86faa474c886411      0t0  TCP localhost:9230 (LISTEN)
node      19207 chris   45u  IPv4 0xb86faa474c8ad411      0t0  TCP localhost:9230->localhost:63416 (ESTABLISHED)

```

Now I have a second, adjacent, problem. The DevTools Sources panel doesn't list any files for inspection. If I `cmd+P` and type in my application name, I can open up e.g. `_app.tsx` for inspection, and `index.tsx`, but I can't find any request/response handlers. Searching for them by name does not yield any relevant results. How do I find these files? 

I'm grateful for any help. Thank you!

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.6.0: Mon Dec 19 20:44:01 PST 2022; root:xnu-8020.240.18~2/RELEASE_X86_64
    Binaries:
      Node: 19.5.0
      npm: 9.3.1
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.2.4-canary.4
      eslint-config-next: 13.2.3
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by emmanuelbuah** [comment]
I have the same issue

**Comment by greatvivek11** [comment]
This is working for both client and server-side code. I'm using Next 14.1.0. After busting my brain for 2 days, found the answer on https://github.com/vercel/next.js/issues/47561#issuecomment-1549241442.
```
{
            "name": "Launch Next.js app",
            "request": "launch",
            "type": "node",
            "runtimeArgs": [
                "--inspect"
            ],
            "program": "${workspaceFolder}/node_modules/.bin/next",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "serverReadyAction": {
                "action": "debugWithEdge",
                "killOnServerStop": true,
                "webRoot": "${workspaceFolder}",
                "pattern": "- Local:.+(https?://.+)",
                "uriFormat": "%s"
            }
        }
```

**Reply by chrisfillmore** [reply]
I'm not familiar with VSCode launch configurations. How does that differ from the below?
```
node --inspect ./node_modules/next/dist/.bin/next dev
```
I acknowledge that this works, but it starts the debugger on port 9230 instead of 9229.

**Reply by greatvivek11** [reply]
Honestly, I've no clue what's going on here. I've been busting my head over this for over 2 days now and finally I got it working but this should be built into NextJS ecosystem. I wonder if Next team itself uses any debugging tools or not?

I read somewhere in one of these issues that next starts 2 ports one for browser and one for server. 9230 is for server as mentioned in the screenshot.

It seems as if either VScode or Chrome debugger locks the port 9230 that's why its not available for Next to bind for setting breakpoints?

Related issues -
https://github.com/vercel/next.js/issues/45697
https://github.com/vercel/next.js/issues/55862
https://github.com/vercel/next.js/issues/47561
https://github.com/vercel/next.js/issues/22835
https://github.com/vercel/next.js/issues/47083

<img width="802" alt="Screenshot 2024-04-11 at 8 51 57 PM" src="https://github.com/vercel/next.js/assets/54403368/aeee627b-7d98-44a6-b6c8-cc735485000d">

package.json
` "dev": "node --inspect ./node_modules/.bin/next dev --turbo",`

**Reply by greatvivek11** [reply]
FYI - .vscode/launch.json 
You will need Microsoft Edge Tools extension else you can replace `debugWithEdge` to `debugWithChrome`. 
https://marketplace.visualstudio.com/items?itemName=ms-edgedevtools.vscode-edge-devtools
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "msedge",
            "name": "Launch Microsoft Edge",
            "request": "launch",
            "cleanUp": "wholeBrowser",
            "cascadeTerminateToConfigurations": [
                "Launch Next.js app",
                "Open Edge DevTools"
            ],
            "url": "http:localhost:3000",
            "presentation": {
                "hidden": true
            }
        },
        {
            "name": "Next.js: debug full stack - 1",
            "request": "launch",
            "type": "node",
            "runtimeArgs": [
                "--inspect"
            ],
            "program": "${workspaceFolder}/node_modules/.bin/next",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "serverReadyAction": {
                "action": "debugWithEdge",
                "killOnServerStop": true,
                "webRoot": "${workspaceFolder}",
                "pattern": "- Local:.+(https?://.+)",
                "uriFormat": "%s"
            }
        },
        {
            "name": "Next.js: debug full stack - 2",
            "type": "node",
            "request": "launch",
            "runtimeArgs": [
                "run",
                "dev" // "next dev --turbo",
            ],
            "runtimeExecutable": "npm",
            "sourceMaps": false,
            "skipFiles": [
                "<node_internals>/**"
            ],
            "internalConsoleOptions": "openOnSessionStart",
            "autoAttachChildProcesses": true,
            "presentation": {
                "hidden": false
            },
            "serverReadyAction": {
                "action": "debugWithEdge",
                "pattern": "- Local:.+(https?://.+)",
                "uriFormat": "%s"
            }
        },
        {
            "type": "msedge",
            "name": "Launch Microsoft Edge in headless mode",
            "request": "launch",
            "outFiles": [
                "${workspaceFolder}/**/*.(m|c|)js",
                "!**/node_modules/**"
            ],
            "resolveSourceMapLocations": [
                "!**/node_modules/**",
                "!**/.next/**",
                "/bin/**"
            ],
            "sourceMaps": true,
            "runtimeArgs": [
                "--headless",
                "--remote-debugging-port=9222"
            ],
            "url": "http:localhost:3000",
            "presentation": {
                "hidden": true
            }
        },
        {
            "type": "vscode-edge-devtools.debug",
            "name": "Open Edge DevTools",
            "request": "attach",
            "url": "http:localhost:3000",
            "presentation": {
                "hidden": true
            }
        }
    ],
    "compounds": [
        {
            "name": "Launch Edge Headless and attach DevTools",
            "configurations": [
                "Launch Microsoft Edge in headless mode",
                "Open Edge DevTools",
                "Next.js: debug full stack"
            ],
            "stopAll": true
        },
        {
            "name": "Launch Edge and attach DevTools",
            "configurations": [
                "Launch Microsoft Edge",
                "Open Edge DevTools",
                "Launch Next.js app"
            ],
            "stopAll": true
        }
    ]
}
```
